src/wh_server_dma.c

```diff
@@ -0,0 +1,307 @@
+#include "wolfhsm/wh_server.h"
+#include "wolfhsm/wh_error.h"
+#include "wolfhsm/wh_server_dma.h"
+
+/* TODO: if the Address allowlist ever gets large, we should consider a more
+ * efficient representation (requiring sorted array and binary search, or
+ * building a binary tree, etc.) */
+
+
+static int _checkAddrAgainstAllowList(const whServerDmaAddrList allowList, void* addr,
+                                      size_t size)
+{
+    uintptr_t startAddr = (uintptr_t)addr;
+    uintptr_t endAddr   = startAddr + size;
+
+    /* Check if the address range is fully within a allowlist entry */
+    for (int i = 0; i < WH_DMA_ADDR_ALLOWLIST_SIZE; i++) {
+        uintptr_t allowlistStartAddr = (uintptr_t)allowList[i].addr;
+        uintptr_t allowlistEndAddr   = allowlistStartAddr + allowList[i].size;
+
+        if (startAddr >= allowlistStartAddr && endAddr <= allowlistEndAddr) {
+            return WH_ERROR_OK;
+        }
+    }
+
+    return WH_ERROR_ACCESS;
+}
+
+static int _checkMemOperAgainstAllowList(const whServerDmaAddrAllowList* allowList,
+                                         whServerDmaOper oper, void* addr,
+                                         size_t size)
+{
+    int rc = WH_ERROR_OK;
+
+    /* If a read/write operation is requested, check the transformed address
+     * against the appropriate allowlist
+     *
+     * TODO: do we need to allowlist check on POST in case there are subsequent
+     * memory operations for some reason?
+     */
+    if (oper == WH_DMA_OPER_CLIENT_READ_PRE) {
+        rc = _checkAddrAgainstAllowList(allowList->readList, addr, size);
+    }
+    else if (oper == WH_DMA_OPER_CLIENT_WRITE_PRE) {
+        rc = _checkAddrAgainstAllowList(allowList->writeList, addr, size);
+    }
+
+    return rc;
+}
+
+int wh_Server_DmaRegisterCb32(whServerContext* server, whServerDmaClientMem32Cb cb)
+{
+    if (NULL == server || NULL == cb) {
```

Comment:
Should cb==NULL be used to use a dummy/default CB that allows everything and does no translation?

Reply:
It already works like this, albeit without calling a "default callback" (it just doesn't call a callback if NULL), and I don't allow you to explicitly register a NULL callback using this function. When initializing the server config struct, if the callback is left NULL that is exactly how it will behave. I think if we ever have logic that a default callback would need to do on every target that doesn't register one, we can definitely add it in, but right now it just does nothing.